### ISSUE TITLE
Possible type inference regression (from v3.2.4 onwards)

### ISSUE THREAD

**Issue by tlaitinen** [microsoft/TypeScript#29914]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:** type inference regression

**Code**

```ts
export function f<A>() {
  let a:undefined extends A | undefined ? A | undefined : undefined = undefined;
  return a;
}
```

**Expected behavior:** No type error. TypeScript compiler version 3.2.2 gives no error.

**Actual behavior:** 
Starting version 3.2.4 TypeScript compiler gives the following type error:
```
$ tsc --strict error.ts
error.ts:2:7 - error TS2322: Type 'undefined' is not assignable to type 'undefined extends A | undefined ? A | undefined : undefined'.

2   let a:undefined extends A | undefined ? A | undefined : undefined = undefined;
```

**Playground Link:**  could not reproduce in TypeScript Playground. 

**Related Issues:**

**Comment by jack-williams** [comment]
Looks related to #29662

**Comment by DanielRosenwasser** [comment]
Looks to be fixed in 4.3.

```ts repro
export function f<A>() {
    let a: undefined extends A | undefined ? A | undefined : undefined = undefined;
    return a;
}
```
